misk-aws/src/main/kotlin/misk/jobqueue/sqs/SqsTransactionalJobQueue.kt

```diff
@@ -0,0 +1,46 @@
+package misk.jobqueue.sqs
+
+import misk.hibernate.Gid
+import misk.hibernate.Session
+import misk.jobqueue.JobQueue
+import misk.jobqueue.QueueName
+import misk.jobqueue.TransactionalJobQueue
+import misk.logging.getLogger
+import java.time.Duration
+import javax.inject.Inject
+import javax.inject.Singleton
+
+// Implements the TransactionalJobQueue interface by registering a post-commit hook that
+// forwards the message to SQS. It is not truly transactional, and is intended only for
+// development / staging while we build proper job forwarding from the replication stream,
```

Comment:
For it to be actually transactional (in particular, atomic + durable wrt the local database transaction) you always have to write to a local spool. We could then attempt an enqueue followed by a delete, but then you still run into the issue that the replication stream observer might see the insert before you finished the enqueue+delete. So I think once we decide to use replication stream as a forwarding system, it's kindof all or nothing

Reply:
Jobsy had this two phase enqueue and now that I think about it I think we can use it here too.

The first phase we just enqueue directly to Jobsy and I think we even did that inside the transaction. We then also did a transaction local write to a "ready ticket" table inside the transaction. On the processing side we checked the ready ticket table to see if the transaction that posted actually committed. If it didn't commit (yet) we don't ack the message and retry for a configurable amount of time (until the transaction is assumed to have rolled back).

(We also did an optimized "happy path" which cached the ready ticket in Redis no a post commit hook. If the ready ticket wasn't in Redis we had to fall back to checking the ready ticket table. This dramatically drops qps on the ready ticket table.)

The advantage is that we don't need a polling job or streaming the binlogs.

The main drawback is that we call out to Jobsy inside the transaction so you gotta have a pretty short timeout on that call. I don't believe we've had issues with this in Franklin.

Of course this only works if the sender and receiver of a job is the same application so that they can both read and write the ready ticket table (which should always be the case for job queueing but won't work for a pubsub).

Reply:
Yeah I explicitly rejected the jobsy approach of doing an RPC with the database transaction open. We get away with this in datacenter because Franklin and jobsy are quite close so the median latency is very low and the latency distribution is pretty tight. Neither of these hold for SQS. p50 on an SQS send is ~100ms and the range of latencies is much broader. Even though median is well below the timeout, adding an extra 100-200ms to every database transaction is going to increase lock contention and cause transactions to pile up (as we saw when we had the few hundred extra ms of cross-shard latency awhile back). 

Besides, the jobsy model won't work for evently and we need to deliver real transactional publishes for both at around the same time.